external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/FileLock.java

```diff
@@ -0,0 +1,328 @@
+    long now = System.currentTimeMillis();
+    LogEntry entry = new LogEntry(now, componentID, fileOffset);
+    String line = entry.toString();
+    if(prefixNewLine) {
+      lockFileStream.writeBytes(System.lineSeparator() + line);
+    }
+    else {
+      lockFileStream.writeBytes(line);
+    }
+    lockFileStream.hflush();
+
+    lastEntry = entry; // update this only after writing to hdfs
+  }
+
+  /** Release lock by deleting file
+   * @throws IOException if lock file could not be deleted
+   */
+  public void release() throws IOException {
+    lockFileStream.close();
+    if(!fs.delete(lockFile, false)) {
+      LOG.warn("Unable to delete lock file, Spout = {}", componentID);
+      throw new IOException("Unable to delete lock file");
+    }
+    LOG.debug("Released lock file {}. Spout {}", lockFile, componentID);
+  }
+
+  // For testing only.. invoked via reflection
+  private void forceCloseLockFile() throws IOException {
+    lockFileStream.close();
+  }
+
+  /** returns lock on file or null if file is already locked. throws if unexpected problem */
+  public static FileLock tryLock(FileSystem fs, Path fileToLock, Path lockDirPath, String spoutId)
+          throws IOException {
+    String lockFileName = lockDirPath.toString() + Path.SEPARATOR_CHAR + fileToLock.getName();
+    Path lockFile = new Path(lockFileName);
+
+    try {
+      FSDataOutputStream ostream = HdfsUtils.tryCreateFile(fs, lockFile);
+      if (ostream != null) {
+        LOG.debug("Acquired lock on file {}. LockFile= {}, Spout = {}", fileToLock, lockFile, spoutId);
+        return new FileLock(fs, lockFile, ostream, spoutId);
+      } else {
+        LOG.debug("Cannot lock file {} as its already locked. Spout = {}", fileToLock, spoutId);
+        return null;
+      }
+    } catch (IOException e) {
+      LOG.error("Error when acquiring lock on file " + fileToLock + " Spout = " + spoutId, e);
+      throw e;
+    }
+  }
+
+  /**
+   * checks if lockFile is older than 'olderThan' UTC time by examining the modification time
+   * on file and (if necessary) the timestamp in last log entry in the file. If its stale, then
+   * returns the last log entry, else returns null.
+   * @param fs
+   * @param lockFile
+   * @param olderThan  time (millis) in UTC.
+   * @return the last entry in the file if its too old. null if last entry is not too old
+   * @throws IOException
+   */
+  public static LogEntry getLastEntryIfStale(FileSystem fs, Path lockFile, long olderThan)
+          throws IOException {
+    long modifiedTime = fs.getFileStatus(lockFile).getModificationTime();
+    if( modifiedTime <= olderThan ) { // look
+      //Impt: HDFS timestamp may not reflect recent appends, so we double check the
+      // timestamp in last line of file to see when the last update was made
+      LogEntry lastEntry =  getLastEntry(fs, lockFile);
+      if(lastEntry==null) {
+        throw new RuntimeException(lockFile.getName() + " is empty. this file is invalid.");
```

Comment:
Thanks for the explanation about when this can happen, but I am more concerned about how we can recover from this.  The RuntimeException when it does happen looks like it will bubble all the way up to nextTuple which will trigger the worker to shut down and restart.  But because nothing has happened to the corrupted file all of the spouts as they finish parsing their file will now go into an endless restart loop.

Can we at least think about how we can recover from this situation.  Or at least not crash constantly when it does happen and yell a lot in the logs.